rmf_site_format/src/workcell.rs

```diff
@@ -289,6 +344,110 @@ impl Workcell {
+            .into_iter()
+            .map(|v| {
+                let visual = v.1.bundle;
+                let visual = urdf_rs::Visual {
+                    name: Some(visual.name),
+                    origin: visual.pose.into(),
+                    geometry: visual.geometry.into(),
+                    material: None,
+                };
+                (visual, v.1.parent)
+            })
+            .collect();
+
+        let collisions_and_parents: Vec<(urdf_rs::Collision, _)> = workcell
+            .collisions
+            .into_iter()
+            .map(|c| {
+                let collision = c.1.bundle;
+                let collision = urdf_rs::Collision {
+                    name: Some(collision.name),
+                    origin: collision.pose.into(),
+                    geometry: collision.geometry.into(),
+                };
+                (collision, c.1.parent)
+            })
+            .collect();
+
+        let links: Result<Vec<_>, WorkcellToUrdfError> = workcell
+            .frames
+            .into_iter()
+            .map(|f| {
+                let frame = f.1.bundle;
+
+                let name = match frame.name {
+                    Some(name) => name.0,
+                    None => format!("frame_{}", f.0),
+                };
+                let visual: Vec<Visual> = visuals_and_parents
+                    .iter()
+                    .filter(|(_, parent)| parent == &f.1.parent)
+                    .map(|(visual, _)| visual.clone())
+                    .collect();
+                let collision: Vec<Collision> = collisions_and_parents
+                    .iter()
+                    .filter(|(_, parent)| parent == &f.1.parent)
+                    .map(|(collision, _)| collision.clone())
+                    .collect();
+
+                let pose: urdf_rs::Pose;
+                if let Anchor::Pose3D(p) = frame.anchor {
+                    pose = p.into();
+                } else {
+                    return Err(WorkcellToUrdfError::InvalidAnchorType(frame.anchor));
+                }
+                let link = urdf_rs::Link {
+                    name,
+                    inertial: urdf_rs::Inertial {
+                        origin: pose,
+                        inertia: {
+                            urdf_rs::Inertia {
+                                ixx: 0.0,
+                                ixy: 0.0,
+                                ixz: 0.0,
+                                iyy: 0.0,
+                                iyz: 0.0,
+                                izz: 0.0,
+                            }
+                        },
+                        mass: urdf_rs::Mass { value: 0.0 },
+                    },
+                    collision,
+                    visual,
+                };
+                Ok(link)
+            })
+            .collect();
+        let links = match links {
+            Ok(links) => links,
+            Err(e) => return Err(e),
+        };
```

Comment:
There is one of my favorite operators `?` it basically "unwraps" the result returning the error if it failed.
You can just apply it to the structure above by removing the type annotation from `link` and changing the `collect()` to:

```
.collect::<Result<Vec<_>, _>>()?;
```

What _I believe_ this does is still communicate the `collect` function to bring the errors in the inner `map` function to the external collection, then unwrap it or return the error.
In this way you can get rid of this match statement entirely.

Reply:
That's awesome - thanks for the tip!